File: docs/imperativo/teste.c
#include <stdio.h>

#define N 9

static int tabuleiro[N][N];

static int posicao_segura(int linha, int coluna, int valor) {
    int i, j, bloco_linha, bloco_coluna;

    for (i = 0; i < N; i++) {
        if (tabuleiro[linha][i] == valor) return 0;   
        if (tabuleiro[i][coluna] == valor) return 0;
    }

    bloco_linha = (linha / 3) * 3;
    bloco_coluna = (coluna / 3) * 3;
    for (i = 0; i < 3; i++) {
        for (j = 0; j < 3; j++) {
            if (tabuleiro[bloco_linha + i][bloco_coluna + j] == valor) return 0;
        }
    }
    return 1;
}

static int entrada_valida(void) {
    int linha, coluna, valor;
    for (linha = 0; linha < N; linha++) {
        for (coluna = 0; coluna < N; coluna++) {
            valor = tabuleiro[linha][coluna];
            if (valor != 0) {
                tabuleiro[linha][coluna] = 0;
                if (!posicao_segura(linha, coluna, valor)) {
                    tabuleiro[linha][coluna] = valor;
                    return 0;
                }
                tabuleiro[linha][coluna] = valor;
            }
        }
    }
    return 1;
}

static int resolver(void) {
    int linha, coluna, valor;

    for (linha = 0; linha < N; linha++) {
        for (coluna = 0; coluna < N; coluna++) {
            if (tabuleiro[linha][coluna] == 0) {
                for (valor = 1; valor <= 9; valor++) {
                    if (posicao_segura(linha, coluna, valor)) {
                        tabuleiro[linha][coluna] = valor;
                        if (resolver()) return 1;
                        tabuleiro[linha][coluna] = 0;
                    }
                }
                return 0; 
            }
        }
    }
    return 1;
}

static int ler_entrada(void) {
    char buffer[256];
    int linha, coluna, tamanho;

    for (linha = 0; linha < N; linha++) {
        if (fgets(buffer, sizeof(buffer), stdin) == NULL) {
            return 0;
        }

        tamanho = 0;
        while (buffer[tamanho] != '\0' && buffer[tamanho] != '\n' && buffer[tamanho] != '\r') {
            tamanho++;
        }

        if (tamanho != N) {
            return 0;
        }

        for (coluna = 0; coluna < N; coluna++) {
            if (buffer[coluna] < '0' || buffer[coluna] > '9') {
                return 0;
            }
            tabuleiro[linha][coluna] = buffer[coluna] - '0';
        }
    }
    return 1;
}

static void imprimir_tabuleiro(void) {
    int linha, coluna;
    for (linha = 0; linha < N; linha++) {
        for (coluna = 0; coluna < N; coluna++) {
            printf("%d", tabuleiro[linha][coluna]);
        }
        printf("\n");
    }
}

int main(void) {
    if (!ler_entrada()) {
        printf("ENTRADA INVALIDA\n");
        return 0;
    }

    if (!entrada_valida() || !resolver()) {
        printf("SEM SOLUCAO\n");
        return 0;
    }

    imprimir_tabuleiro();
    return 0;
}

File: docs/poo/teste.java
// [P4-ETAPA-04] Implementacao orientada a objetos
// Paradigma: Orientacao a Objetos
// Linguagem: Java
//
// Contrato (ver docs/especificacao.md):
//   Entrada: 9 linhas de 9 digitos (0 = celula vazia) via stdin.
//   Saida:   9 linhas de 9 digitos (solucao completa) via stdout,
//            ou a linha "SEM SOLUCAO".
//
// Estilo OOP: o estado (o tabuleiro) e' encapsulado numa classe,
// que expoe metodos de consulta e alteracao controlada, em vez de
// arrays manipulados livremente por funcoes externas. A logica de
// resolucao fica separada, atuando sobre a interface do Tabuleiro.

import java.util.Scanner;

/** Representa o tabuleiro do Sudoku e encapsula suas regras de validade. */
class Tabuleiro {
    private static final int N = 9;
    private final int[][] celulas = new int[N][N];

    public int tamanho() {
        return N;
    }

    public int valorEm(int linha, int coluna) {
        return celulas[linha][coluna];
    }

    public boolean vazia(int linha, int coluna) {
        return celulas[linha][coluna] == 0;
    }

    public void definir(int linha, int coluna, int valor) {
        celulas[linha][coluna] = valor;
    }

    public void limpar(int linha, int coluna) {
        celulas[linha][coluna] = 0;
    }

    /** verificação de regras do sudoku */
    public boolean aceitaValor(int linha, int coluna, int valor) {
        for (int i = 0; i < N; i++) {
            if (celulas[linha][i] == valor) return false;
            if (celulas[i][coluna] == valor) return false;
        }
        int blocoLinha = (linha / 3) * 3;
        int blocoColuna = (coluna / 3) * 3;
        for (int i = 0; i < 3; i++) {
            for (int j = 0; j < 3; j++) {
                if (celulas[blocoLinha + i][blocoColuna + j] == valor) return false;
            }
        }
        return true;
    }

    /** Confirma que as celulas ja preenchidas nao se contradizem entre si. */
    public boolean entradaValida() {
        for (int linha = 0; linha < N; linha++) {
            for (int coluna = 0; coluna < N; coluna++) {
                int valor = celulas[linha][coluna];
                if (valor != 0) {
                    celulas[linha][coluna] = 0;
                    boolean aceita = aceitaValor(linha, coluna, valor);
                    celulas[linha][coluna] = valor;
                    if (!aceita) return false;
                }
            }
        }
        return true;
    }

    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder();
        for (int linha = 0; linha < N; linha++) {
            for (int coluna = 0; coluna < N; coluna++) {
                sb.append(celulas[linha][coluna]);
            }
            sb.append('\n');
        }
        return sb.toString();
    }
}

/** interage apenas pela interface publica da classe Tabuleiro. */
class SudokuSolver {
    public boolean resolver(Tabuleiro tab) {
        int n = tab.tamanho();
        for (int linha = 0; linha < n; linha++) {
            for (int coluna = 0; coluna < n; coluna++) {
                if (tab.vazia(linha, coluna)) {
                    for (int valor = 1; valor <= 9; valor++) {
                        if (tab.aceitaValor(linha, coluna, valor)) {
                            tab.definir(linha, coluna, valor);
                            if (resolver(tab)) return true;
                            tab.limpar(linha, coluna); // backtrack
                        }
                    }
                    return false;
                }
            }
        }
        return true;
    }
}

public class Sudoku {
    public static void main(String[] args) {
        Tabuleiro tabuleiro = new Tabuleiro();
        Scanner leitor = new Scanner(System.in);

        for (int linha = 0; linha < tabuleiro.tamanho(); linha++) {
            String linhaTexto = leitor.next();
            for (int coluna = 0; coluna < tabuleiro.tamanho(); coluna++) {
                tabuleiro.definir(linha, coluna, linhaTexto.charAt(coluna) - '0');
            }
        }
        leitor.close();

        SudokuSolver solver = new SudokuSolver();
        if (!tabuleiro.entradaValida() || !solver.resolver(tabuleiro)) {
            System.out.println("SEM SOLUCAO");
            return;
        }

        System.out.print(tabuleiro);
    }
}
